sweep/aggregator.go

```diff
@@ -0,0 +1,349 @@
+		inputClusters = append(inputClusters, inputCluster{
+			lockTime:     &lt,
+			sweepFeeRate: sweepFeeRate,
+			inputs:       cluster,
+		})
+	}
+
+	return inputClusters, rem
+}
+
+// clusterBySweepFeeRate takes the set of pending inputs within the UtxoSweeper
+// and clusters those together with similar fee rates. Each cluster contains a
+// sweep fee rate, which is determined by calculating the average fee rate of
+// all inputs within that cluster.
+func (s *SimpleAggregator) clusterBySweepFeeRate(
+	inputs pendingInputs) []inputCluster {
+
+	bucketInputs := make(map[int]*bucketList)
+	inputFeeRates := make(map[wire.OutPoint]chainfee.SatPerKWeight)
+
+	// First, we'll group together all inputs with similar fee rates. This
+	// is done by determining the fee rate bucket they should belong in.
+	for op, input := range inputs {
+		feeRate, err := input.params.Fee.Estimate(
+			s.FeeEstimator, s.MaxFeeRate,
+		)
+		if err != nil {
+			log.Warnf("Skipping input %v: %v", op, err)
+			continue
+		}
+
+		// Only try to sweep inputs with an unconfirmed parent if the
+		// current sweep fee rate exceeds the parent tx fee rate. This
+		// assumes that such inputs are offered to the sweeper solely
+		// for the purpose of anchoring down the parent tx using cpfp.
+		parentTx := input.UnconfParent()
+		if parentTx != nil {
+			parentFeeRate :=
+				chainfee.SatPerKWeight(parentTx.Fee*1000) /
+					chainfee.SatPerKWeight(parentTx.Weight)
+
+			if parentFeeRate >= feeRate {
+				log.Debugf("Skipping cpfp input %v: "+
+					"fee_rate=%v, parent_fee_rate=%v", op,
+					feeRate, parentFeeRate)
+
+				continue
+			}
+		}
+
+		feeGroup := s.bucketForFeeRate(feeRate)
+
+		// Create a bucket list for this fee rate if there isn't one
+		// yet.
+		buckets, ok := bucketInputs[feeGroup]
+		if !ok {
+			buckets = &bucketList{}
+			bucketInputs[feeGroup] = buckets
+		}
+
+		// Request the bucket list to add this input. The bucket list
+		// will take into account exclusive group constraints.
+		buckets.add(input)
+
+		input.lastFeeRate = feeRate
+		inputFeeRates[op] = feeRate
+	}
+
+	// We'll then determine the sweep fee rate for each set of inputs by
+	// calculating the average fee rate of the inputs within each set.
+	inputClusters := make([]inputCluster, 0, len(bucketInputs))
+	for _, buckets := range bucketInputs {
+		for _, inputs := range buckets.buckets {
+			var sweepFeeRate chainfee.SatPerKWeight
+			for op := range inputs {
+				sweepFeeRate += inputFeeRates[op]
+			}
+			sweepFeeRate /= chainfee.SatPerKWeight(len(inputs))
+			inputClusters = append(inputClusters, inputCluster{
+				sweepFeeRate: sweepFeeRate,
+				inputs:       inputs,
+			})
+		}
+	}
+
+	return inputClusters
+}
+
+// bucketForFeeReate determines the proper bucket for a fee rate. This is done
+// in order to batch inputs with similar fee rates together.
+func (s *SimpleAggregator) bucketForFeeRate(
+	feeRate chainfee.SatPerKWeight) int {
+
+	relayFeeRate := s.FeeEstimator.RelayFeePerKW()
```

Comment:
Since we're no longer using the static `s.relayFeeRate`, this `relayFeeRate` can change between iterations and potentially bucket inputs that shouldn't be bucketed together. I think instead we should call `RelayFeePerKW` once at the top of the calling function and then pass that value in here.

I haven't checked the usage of `s.relayFeeRate` in later PRs yet, but after this PR it's only used in the anchor resolver post-force close and it may be a stale value. Not a huge deal though, but would be good to sweep the anchor with a valid min relay fee.